src/main/java/htsjdk/samtools/util/EdgeReadIterator.java

```diff
@@ -69,13 +75,58 @@ public EdgeReadIterator(final SamReader samReader, final IntervalList intervalLi
      * @param useIndex     If true, do indexed lookup to improve performance.  Not relevant if intervalList == null.
      *                     It is no longer the case the useIndex==true can make performance worse.  It should always perform at least
      *                     as well as useIndex==false, and generally will be much faster.
      */
     public EdgeReadIterator(final SamReader samReader, final IntervalList intervalList, final boolean useIndex) {
-        super(samReader, intervalList, useIndex);
+        super(samReader,
+                intervalList == null || intervalList.getHeader().getSortOrder() == SAMFileHeader.SortOrder.coordinate ? intervalList : intervalList.sorted(),
+                useIndex);
+
+        if (getIntervals() == null) {
+            intervalListIterator = null;
+            intervalCoordinateComparator = null;
+            overlapDetector = null;
+        } else {
+            // For the easy case of a read being fully contained in an interval, we use an iterator to keep track
+            // of the current interval.
+            intervalListIterator = getIntervals().iterator();
+            if (intervalListIterator.hasNext()) {
+                currentInterval = intervalListIterator.next();
+            }
+            if (intervalListIterator.hasNext()) {
+                nextInterval = intervalListIterator.next();
+            }
+            // We also need this comparator to keep track of the different contigs. The constructor of
+            // AbstractLocusInfo ensures that the sequence dictionaries of the SAM file and the interval list match.
+            intervalCoordinateComparator = new IntervalCoordinateComparator(getHeader());
+
+            // For more complicated cases, we need an OverlapDetector.
+            overlapDetector = OverlapDetector.create(getIntervals());
+        }
+    }
+
+    /**
+     * This function updates currentInterval and nextInterval according to the position of the record that it is
+     * presented and determines if the current read is fully contained in the currentInterval.
+     * @param rec The record we want to consider
+     * @return True, if rec is fully contained in the current interval, otherwise false
+     */
+     protected boolean intervalCompletelyContainsRead(final SAMRecord rec) {
+         // currentInterval should never be null when calling this method, but we have to check it just to make sure,
+         // so that we don't get a NullPointerException in the return statement.
+         if (currentInterval == null) {
+             return false;
+         }
+        // Here we need to update the currentInterval. We have to do this using an
+        // IntervalCoordinateComparator to take factor in the order in the sequence dictionary.
+        while (nextInterval != null && intervalCoordinateComparator.compare(new Interval(rec), nextInterval) > 0) {
```

Comment:
Another good point (and good catch!). I'm using the IntervalCoordinateComparator mostly for convenience, because it takes into account the sort order of the contigs in the sequence dictionary, because they might not be alphabetically sorted. And you're right, it could happen that having a strand or name specified could affect the comparison, but that happens only after looking at start and end coordinates. So this would only be an issue if the current read happens to have the same start and end coordinates as the interval (which would be very rare), and in such cases I think we can afford that we get this wrong, because worst case we will process a few reads with the slower implementation.